### PR TITLE
Change edit mode + add edit lovelace

### PR DESCRIPTION
**Pull request by bramkragten**
image

### CODE REVIEW

File: src/panels/lovelace/editor/lovelace-editor/hui-dialog-edit-lovelace.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { html, LitElement, PropertyDeclarations } from "@polymer/lit-element";
+import { TemplateResult } from "lit-html";
+import { Lovelace } from "../../types";
+
+import { HomeAssistant } from "../../../../types";
+import "./hui-edit-lovelace";
+
+export class HuiDialogEditLovelace extends LitElement {
```
Comment (balloob) [inline src/panels/lovelace/editor/lovelace-editor/hui-dialog-edit-lovelace.ts]: This class does literally nothing. It passes all received properties and function calls to the element it renders

File: src/panels/lovelace/editor/lovelace-editor/hui-edit-lovelace.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+              alt="Saving"
+            ></paper-spinner>
+            ${this.localize("ui.common.save")}</paper-button
+          >
+        </div>
+      </paper-dialog>
+    `;
+  }
+
+  private renderStyle(): TemplateResult {
```
Comment (balloob) [inline src/panels/lovelace/editor/lovelace-editor/hui-edit-lovelace.ts]: I think that we should make it a habit to move this to the bottom of the file.

File: src/panels/lovelace/editor/lovelace-editor/hui-dialog-edit-lovelace.ts (hunk 1 above)
Comment (bramkragten) [reply src/panels/lovelace/editor/lovelace-editor/hui-dialog-edit-lovelace.ts]: That's what I meant, it has a lot of stuff now that is not needed for how simple it is now....

File: src/panels/lovelace/editor/lovelace-editor/hui-lovelace-editor.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  }
+}
+
+export class HuiLovelaceEditor extends hassLocalizeLitMixin(LitElement) {
+  static get properties(): PropertyDeclarations {
+    return { hass: {}, _config: {} };
+  }
+
+  get _title(): string {
+    if (!this._config) {
```
Comment (balloob) [inline src/panels/lovelace/editor/lovelace-editor/hui-lovelace-editor.ts]: is this possible?

File: src/panels/lovelace/editor/lovelace-editor/hui-lovelace-editor.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+  public hass?: HomeAssistant;
+  private _config?: LovelaceConfig;
+
+  set config(config: LovelaceConfig) {
+    this._config = config;
+  }
+
+  protected render(): TemplateResult {
+    if (!this.hass) {
```
Comment (balloob) [inline src/panels/lovelace/editor/lovelace-editor/hui-lovelace-editor.ts]: Why do we need hass ?

File: src/panels/lovelace/editor/lovelace-editor/hui-lovelace-editor.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      return html``;
+    }
+
+    return html`
+      ${configElementStyle}
+      <div class="card-config">
+        <paper-input
+          label="Title"
+          value="${this._title}"
```
Comment (balloob) [inline src/panels/lovelace/editor/lovelace-editor/hui-lovelace-editor.ts]: `.value`

File: src/panels/lovelace/editor/lovelace-editor/hui-lovelace-editor.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    let newConfig;
+
+    if (target.configValue) {
+      newConfig = {
+        ...this._config,
+        [target.configValue]: target.value,
+      };
+    }
+
+    fireEvent(this, "lovelace-config-changed", { config: newConfig });
```
Comment (balloob) [inline src/panels/lovelace/editor/lovelace-editor/hui-lovelace-editor.ts]: I think that we should make these events not bubble. That way only the parent can listen to it and that would make it faster?

Guess it doesn't really matter in the grand scheme of things.

File: src/panels/lovelace/editor/lovelace-editor/show-edit-lovelace-dialog.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+let registeredDialog = false;
+const dialogShowEvent = "show-edit-lovelace";
+const dialogTag = "hui-dialog-edit-lovelace";
+
+const registerEditLovelaceDialog = (element: HTMLElement) =>
+  fireEvent(element, "register-dialog", {
+    dialogShowEvent,
+    dialogTag,
+    dialogImport: () => import("./hui-dialog-edit-lovelace"),
```
Comment (balloob) [inline src/panels/lovelace/editor/lovelace-editor/show-edit-lovelace-dialog.ts]: Name this chunk. See #2278

File: src/panels/lovelace/hui-root.ts [hunk 8]
```diff
@@ -120,18 +121,51 @@ class HUIRoot extends hassLocalizeLitMixin(LitElement) {
...
+                    <paper-icon-button
+                      icon="hass:dots-vertical"
+                      slot="dropdown-trigger"
+                    ></paper-icon-button>
+                    <paper-listbox
+                      @iron-select="${this._deselect}"
+                      slot="dropdown-content"
+                    >
+                      <paper-item @click="${this._editModeDisable}"
+                        >Close UI editor</paper-item
```
Comment (balloob) [inline src/panels/lovelace/hui-root.ts]: No need for this one since we have close on the left.

File: src/panels/lovelace/hui-root.ts [hunk 9]
```diff
@@ -120,18 +121,51 @@ class HUIRoot extends hassLocalizeLitMixin(LitElement) {
...
+                      >
+                      <paper-item @click="${this.lovelace!.enableFullEditMode}"
+                        >Raw config editor</paper-item
+                      >
+                      <paper-item>Add custom resources</paper-item>
```
Comment (balloob) [inline src/panels/lovelace/hui-root.ts]: Doesn't do anything yet.

File: src/panels/lovelace/hui-root.ts [hunk 10]
```diff
@@ -120,18 +121,51 @@ class HUIRoot extends hassLocalizeLitMixin(LitElement) {
...
+                    ></paper-icon-button>
+                    <paper-listbox
+                      @iron-select="${this._deselect}"
+                      slot="dropdown-content"
+                    >
+                      <paper-item @click="${this._editModeDisable}"
+                        >Close UI editor</paper-item
+                      >
+                      <paper-item @click="${this.lovelace!.enableFullEditMode}"
+                        >Raw config editor</paper-item
```
Comment (balloob) [inline src/panels/lovelace/hui-root.ts]: Good to add it here, we should remove it from the normal dropdown.

File: src/panels/lovelace/hui-root.ts [hunk 11]
```diff
@@ -120,18 +121,51 @@ class HUIRoot extends hassLocalizeLitMixin(LitElement) {
...
+                      >
+                      <paper-item>Add custom resources</paper-item>
+                      <paper-item @click="${this._handleHelp}">Help</paper-item>
```
Comment (balloob) [inline src/panels/lovelace/hui-root.ts]: We should move help to a paper-icon-button that lives next to the menu.

Comment (bramkragten) [reply src/panels/lovelace/hui-root.ts]: Is that for both views or only edit?

**Comment by bramkragten** [comment]
I don't think we should do that as it would prevent YAML mode users from accessing it...

**Comment by balloob** [comment]
Good point. Freaking YAML users!

**Pull request merged**